File: apps/windows/Auth/SupabaseProfileClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json.Serialization;

namespace Andivum_Windows.Auth;

public sealed class SupabaseProfileClient
{
    private readonly HttpClient httpClient;
    private readonly ITokenStore tokenStore;
    private readonly Uri profileEndpoint;
    private readonly string publishableKey;

    public SupabaseProfileClient(
        HttpClient httpClient,
        ITokenStore tokenStore,
        Uri supabaseUrl,
        string publishableKey)
    {
        this.httpClient = httpClient;
        this.tokenStore = tokenStore;
        this.publishableKey = publishableKey;
        profileEndpoint = new Uri(
            $"{supabaseUrl.AbsoluteUri.TrimEnd('/')}/rest/v1/app_profiles?select=id,user_id&limit=1");
    }

    public async Task<SessionResponse> GetCurrentSessionAsync(
        CancellationToken cancellationToken = default)
    {
        var token = tokenStore.Read() ??
            throw new InvalidOperationException("No saved session is available.");
        if (string.IsNullOrWhiteSpace(token.AccessToken))
        {
            throw new InvalidOperationException("Supabase access token is required.");
        }

        using var response = await SendProfileRequestAsync(
            HttpMethod.Get,
            token.AccessToken,
            content: null,
            cancellationToken);
        if (response.StatusCode == HttpStatusCode.Unauthorized)
        {
            throw new SessionUnauthorizedException();
        }

        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                $"Supabase profile request failed with HTTP {(int)response.StatusCode}.");
        }

        var profiles = await response.Content.ReadFromJsonAsync<
            IReadOnlyList<SupabaseProfileRow>>(cancellationToken: cancellationToken);
        if (profiles is { Count: > 0 })
        {
            return ToSession(profiles[0]);
        }

        using var created = await SendProfileRequestAsync(
            HttpMethod.Post,
            token.AccessToken,
            new StringContent("{}", Encoding.UTF8, "application/json"),
            cancellationToken);
        if (created.StatusCode == HttpStatusCode.Conflict)
        {
            return await GetCurrentSessionAsync(cancellationToken);
        }

        if (!created.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                $"Supabase profile bootstrap failed with HTTP {(int)created.StatusCode}.");
        }

        var createdProfiles = await created.Content.ReadFromJsonAsync<
            IReadOnlyList<SupabaseProfileRow>>(cancellationToken: cancellationToken);
        var profile = createdProfiles is { Count: > 0 }
            ? createdProfiles[0]
            : throw new InvalidOperationException(
                "Supabase profile bootstrap returned no profile.");
        return ToSession(profile);
    }

    private async Task<HttpResponseMessage> SendProfileRequestAsync(
        HttpMethod method,
        string accessToken,
        HttpContent? content,
        CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, profileEndpoint)
        {
            Content = content,
        };
        request.Headers.Add("apikey", publishableKey);
        request.Headers.Authorization = new AuthenticationHeaderValue(
            "Bearer",
            accessToken);
        if (method == HttpMethod.Post)
        {
            request.Headers.Add("Prefer", "return=representation");
        }

        // HttpClient.SendAsync owns the response; the request can be disposed
        // immediately after the send completes.
        return await httpClient.SendAsync(request, cancellationToken);
    }

    private static SessionResponse ToSession(SupabaseProfileRow profile)
    {
        if (string.IsNullOrWhiteSpace(profile.Id) ||
            string.IsNullOrWhiteSpace(profile.UserId))
        {
            throw new InvalidOperationException(
                "Supabase profile returned no stable user id.");
        }

        return new SessionResponse(profile.UserId, Authenticated: true);
    }

    private sealed record SupabaseProfileRow(
        [property: JsonPropertyName("id")] string? Id,
        [property: JsonPropertyName("user_id")] string? UserId);
}

File: apps/android/app/src/main/java/io/github/andivalsei/andivum/SupabaseProfileClient.kt
package io.github.andivalsei.andivum

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONArray

class SupabaseProfileClient(
    private val supabaseUrl: String,
    private val publishableKey: String,
    private val openConnection: (URL) -> HttpURLConnection = { url ->
        url.openConnection() as HttpURLConnection
    },
) {
    fun getCurrentSession(idToken: String): SessionResponse {
        require(idToken.isNotBlank()) { "Auth0 ID token is required." }

        val profileEndpoint =
            "${supabaseUrl.trimEnd('/')}/rest/v1/app_profiles?select=id&limit=1"
        val existing = request(profileEndpoint, "GET", idToken)
        if (existing.statusCode == HttpURLConnection.HTTP_UNAUTHORIZED) {
            throw SessionUnauthorizedException()
        }
        if (existing.statusCode !in 200..299) {
            throw IllegalStateException(
                "Supabase profile request failed with HTTP ${existing.statusCode}.",
            )
        }

        parseFirstProfile(existing.body)?.let { return it }

        val created = request(profileEndpoint, "POST", idToken, "{}")
        if (created.statusCode == HttpURLConnection.HTTP_CONFLICT) {
            return getCurrentSession(idToken)
        }
        if (created.statusCode !in 200..299) {
            throw IllegalStateException(
                "Supabase profile bootstrap failed with HTTP ${created.statusCode}.",
            )
        }

        return parseFirstProfile(created.body)
            ?: throw IllegalStateException(
                "Supabase profile bootstrap returned no profile.",
            )
    }

    private fun request(
        endpoint: String,
        method: String,
        idToken: String,
        body: String? = null,
    ): Response {
        val connection = openConnection(URL(endpoint)).apply {
            requestMethod = method
            setRequestProperty("apikey", publishableKey)
            setRequestProperty("Authorization", "Bearer $idToken")
            connectTimeout = 10_000
            readTimeout = 10_000
            if (method == "POST") {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Prefer", "return=representation")
            }
        }

        try {
            if (body != null) {
                connection.outputStream.use { output ->
                    output.write(body.toByteArray(Charsets.UTF_8))
                }
            }
            val statusCode = connection.responseCode
            val responseStream = if (statusCode in 200..299) {
                connection.inputStream
            } else {
                connection.errorStream
            }
            val responseBody = responseStream?.bufferedReader()?.use { reader ->
                reader.readText()
            }.orEmpty()
            return Response(statusCode, responseBody)
        } finally {
            connection.disconnect()
        }
    }

    private fun parseFirstProfile(body: String): SessionResponse? {
        val profiles = JSONArray(body)
        if (profiles.length() == 0) {
            return null
        }

        val id = profiles.getJSONObject(0).getString("id")
        require(id.isNotBlank()) { "Supabase profile has no stable id." }
        return SessionResponse(id, authenticated = true)
    }

    private data class Response(
        val statusCode: Int,
        val body: String,
    )
}
